main/src/ca/uwaterloo/flix/language/phase/extra/CodeHinter.scala

```diff
@@ -16,349 +17,113 @@
-    */
-  private def visitTrait(trt: TypedAst.Trait)(implicit index: Index): List[CodeHint] = {
-    val uses = index.usesOf(trt.sym)
-    val isDeprecated = trt.ann.isDeprecated
-    val deprecated = if (isDeprecated) uses.map(CodeHint.Deprecated.apply) else Nil
-    val isExperimental = trt.ann.isExperimental
-    val experimental = if (isExperimental) uses.map(CodeHint.Experimental.apply) else Nil
-    (deprecated ++ experimental).toList
-  }
-
-  /**
-    * Computes code quality hints for the given definition `def0`.
-    */
-  private def visitDef(def0: TypedAst.Def)(implicit root: Root, flix: Flix): List[CodeHint] = {
-    visitExp(def0.exp)
-  }
-
-  /**
-    * Computes code quality hints for the given expression `exp0`.
-    */
-  private def visitExp(exp0: Expr)(implicit root: Root, flix: Flix): List[CodeHint] = exp0 match {
-    case Expr.Var(_, _, _) => Nil
-
-    case Expr.Hole(_, _, _, _) => Nil
-
-    case Expr.HoleWithExp(exp, _, _, _) => visitExp(exp)
-
-    case Expr.OpenAs(_, exp, _, _) => visitExp(exp)
-
-    case Expr.Use(_, _, exp, _) => visitExp(exp)
-
-    case Expr.Cst(_, _, _) => Nil
-
-    case Expr.Lambda(_, exp, _, _) =>
-      visitExp(exp)
-
-    case Expr.ApplyClo(exp1, exp2, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2)
-
-    case Expr.ApplyDef(DefSymUse(sym, loc1), exps, _, _, _, _) =>
-      val hints0 = exps.flatMap(e => checkEffect(sym, e.tpe, e.loc))
-      val hints1 = checkDeprecated(sym, loc1) ++
-        checkExperimental(sym, loc1) ++
-        checkParallel(sym, loc1) ++
-        checkLazy(sym, loc1)
-      hints0 ++ hints1 ++ visitExps(exps)
-
-    case Expr.ApplyLocalDef(_, exps, _, _, _, _) =>
-      visitExps(exps)
-
-    case Expr.ApplySig(_, exps, _, _, _, _) =>
-      visitExps(exps)
-
-    case Expr.Unary(_, exp, _, _, _) =>
-      visitExp(exp)
-
-    case Expr.Binary(_, exp1, exp2, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2)
-
-    case Expr.Let(_, exp1, exp2, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2)
-
-    case Expr.LocalDef(_, _, exp1, exp2, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2)
-
-    case Expr.Region(_, _) => Nil
+  def run(root: TypedAst.Root, sources: Set[String]): List[CodeHint] = {
+    var hints: List[CodeHint] = Nil
 
-    case Expr.Scope(_, _, exp, _, _, _) =>
-      visitExp(exp)
-
-    case Expr.IfThenElse(exp1, exp2, exp3, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2) ++ visitExp(exp3)
-
-    case Expr.Stm(exp1, exp2, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2)
-
-    case Expr.Discard(exp, _, _) =>
-      visitExp(exp)
-
-    case Expr.Match(matchExp, rules, _, _, _) =>
-      visitExp(matchExp) ++ rules.flatMap {
-        case MatchRule(_, guard, exp) => guard.toList.flatMap(visitExp) ::: visitExp(exp)
+    object HintConsumer extends Consumer {
+      override def consumeTraitSymUse(symUse: SymUse.TraitSymUse): Unit = {
+        hints = hints ++ considerTrait(root, symUse.sym, symUse.loc)
```

Comment:
Why interleave different concerns?

You could compute uses separate (in a helper function) and just use that

Reply:
Instead now your Consumers are producing hints. Seems like the opposite of separation of concerns.

Reply:
Doing so would mean "throwing away" the type information we have at hand. When creating hints from these uses, we'd have to then pattern match to regain that information to find the def, enum or trait in question to gain it's annotation.

Ideas 
1. different collections for defs, traits, enums and def calls
2. one stack of `(Annotation, SourceLocation)` and treat all the same (check for each if it has an annotation). Separate stack for def call info `(Symbol, Type, SourceLocation)`

Reply:
i agree that more separation would be good

Reply:
You need to collect uses right? Can you not write a visitors that collects those uses in separate maps without worrying that downstream they will be used for hints?

Reply:
What I could do is collect every occurrence of every trait, enum, def use and def call, each in their own type of collection and then afterwards "consider" each. You mean something like that?

Reply:
Yes

Reply:
How about now?